cores/esp32/esp32-hal-uart.c

```diff
@@ -732,37 +703,27 @@ uartDetectBaudrate(uart_t *uart)
     }
 
     uart_dev_t *hw = UART_LL_GET_HW(uart->num);
-    hw->auto_baud.en = 0;
+    uart_ll_set_autobaud_en(hw, false);
 
     uartStateDetectingBaudrate = false; // Initialize for the next round
-
-    unsigned long baudrate = getApbFrequency() / divisor;
     
-    //log_i("APB_FREQ = %d\nraw baudrate detected = %d", getApbFrequency(), baudrate);
-
-    static const unsigned long default_rates[] = {300, 600, 1200, 2400, 4800, 9600, 19200, 38400, 57600, 74880, 115200, 230400, 256000, 460800, 921600, 1843200, 3686400};
-
+#if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32S2
+    unsigned long baudrate = getApbFrequency() / divisor;  // ESP32 and S2 APB Freq
+#else
+    unsigned long baudrate = (getXtalFrequencyMhz() * 1000000) / divisor;  // S3 and C3 XTAL Frequency
+#endif
```

Comment:
What do you think of doing something like:

```suggestion
#if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32S2
    unsigned long baudrate = getApbFrequency() / divisor;  // ESP32 and S2 APB Freq
#elif CONFIG_IDF_TARGET_ESP32S3 || CONFIG_IDF_TARGET_ESP32C3
    unsigned long baudrate = (getXtalFrequencyMhz() * 1000000) / divisor;  // S3 and C3 XTAL Frequency
#else
    log_e("Baud rate detection not supported.");
    return 0;
#endif
```

Reply:
This is for Arduino 2.0.x - there no other SoC beside ESP32, ESP32-2, ESP32-S3 and ESP32-C3.
The code works the same way for S3 and C3. It was already working fine for the ESP32 and S2.